pkg/volume/csi/csi_plugin_test.go

```diff
@@ -655,93 +695,115 @@ func TestPluginNewMounterWithInline(t *testing.T) {
 			spec:          volume.NewSpecFromPersistentVolume(makeTestPV("test-pv1", 20, testDriver, testVol), true),
 			podUID:        types.UID(fmt.Sprintf("%08X", rand.Uint64())),
 			namespace:     "test-ns1",
-			csiVolumeMode: persistentVolumeMode,
+			csiVolumeMode: storagev1beta1.PersistentVolumeMode,
 		},
 		{
 			name:          "mounter with volume source",
 			spec:          volume.NewSpecFromVolume(makeTestVol("test-vol1", testDriver)),
 			podUID:        types.UID(fmt.Sprintf("%08X", rand.Uint64())),
 			namespace:     "test-ns2",
-			csiVolumeMode: ephemeralVolumeMode,
+			csiVolumeMode: storagev1beta1.EphemeralVolumeMode,
 		},
 	}
 
-	for _, test := range tests {
-		t.Run(test.name, func(t *testing.T) {
-			plug, tmpDir := newTestPlugin(t, nil)
-			defer os.RemoveAll(tmpDir)
+	runAll := func(t *testing.T, supported []storagev1beta1.CSIVolumeMode) {
+		for _, test := range tests {
+			t.Run(test.name, func(t *testing.T) {
+				driver := getTestCSIDriver(testDriver, nil, nil, supported)
+				fakeClient := fakeclient.NewSimpleClientset(driver)
+				plug, tmpDir := newTestPlugin(t, fakeClient)
+				defer os.RemoveAll(tmpDir)
 
-			registerFakePlugin(testDriver, "endpoint", []string{"1.2.0"}, t)
-			mounter, err := plug.NewMounter(
-				test.spec,
-				&api.Pod{ObjectMeta: meta.ObjectMeta{UID: test.podUID, Namespace: test.namespace}},
-				volume.VolumeOptions{},
-			)
-			if test.shouldFail != (err != nil) {
-				t.Fatal("Unexpected error:", err)
-			}
-			if test.shouldFail && err != nil {
-				t.Log(err)
-				return
-			}
+				registerFakePlugin(testDriver, "endpoint", []string{"1.2.0"}, t)
 
-			if mounter == nil {
-				t.Fatal("failed to create CSI mounter")
-			}
-			csiMounter := mounter.(*csiMountMgr)
+				mounter, err := plug.NewMounter(
+					test.spec,
+					&api.Pod{ObjectMeta: meta.ObjectMeta{UID: test.podUID, Namespace: test.namespace}},
+					volume.VolumeOptions{},
+				)
 
-			// validate mounter fields
-			if string(csiMounter.driverName) != testDriver {
-				t.Error("mounter driver name not set")
-			}
-			if csiMounter.volumeID == "" {
-				t.Error("mounter volume id not set")
-			}
-			if csiMounter.pod == nil {
-				t.Error("mounter pod not set")
-			}
-			if string(csiMounter.podUID) != string(test.podUID) {
-				t.Error("mounter podUID not set")
-			}
-			csiClient, err := csiMounter.csiClientGetter.Get()
-			if csiClient == nil {
-				t.Error("mounter csiClient is nil")
-			}
-			if csiMounter.csiVolumeMode != test.csiVolumeMode {
-				t.Error("unexpected driver mode:", csiMounter.csiVolumeMode)
-			}
+				// Some test cases are broken.
+				shouldFail := test.shouldFail
+				// Others fail if the driver does not support the volume mode.
+				if !containsVolumeMode(supported, test.csiVolumeMode) {
+					shouldFail = true
+				}
+				if shouldFail != (err != nil) {
+					t.Fatal("Unexpected error:", err)
+				}
+				if shouldFail && err != nil {
+					t.Log(err)
```

Comment:
Do we need to log? This may be confusing to see in the test logs since this is the success case.

Reply:
This isn't from me. The original code is from `TestPluginNewMounter` (
https://github.com/kubernetes/kubernetes/blob/93bc57e726f74f354b2d97a0d61f9a8310e636b7/pkg/volume/csi/csi_plugin_test.go#L596) but several other tests also do the same.

I prefer to keep it also in the new test, for the sake of consistency.